### PR TITLE
Delegate transaction

### PR DESCRIPTION
**Pull request by alex-arugon**

### CODE REVIEW

File: core/service/transaction/delegate.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        };
+
+        if (asset.username) {
+            asset.username = asset.username.toLowerCase().trim();
+        }
+        return asset;
+    }
+
+    getBytes(trs: Transaction<IAssetDelegate>): Buffer {
+        return Buffer.from(trs.asset.username, 'utf8');
```
Comment (axelsheva) [inline core/service/transaction/delegate.ts]: > Buffer(null, 'utf8');
TypeError [ERR_INVALID_ARG_TYPE]: The first argument must be one of type string, Buffer, ArrayBuffer, Array, or Array-like Object. Received type null
    at Function.from (buffer.js:195:11)
    at Buffer (buffer.js:170:17)

Comment (axelsheva) [reply core/service/transaction/delegate.ts]: Can better make the username required?

File: core/service/transaction/delegate.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+        if ((!trs.asset || !trs.asset.username) && config.constants.VERIFY_DELEGATE_TRS_RECIPIENT) {
+            errors.push('Invalid transaction asset');
+        }
+
+        if (trs.asset.username !== trs.asset.username.toLowerCase() && config.constants.VERIFY_DELEGATE_TRS_RECIPIENT) {
+            errors.push('Username must be lowercase');
+        }
+
+        const isAddress = /^(DDK)+[0-9]{1,25}$/ig;
```
Comment (axelsheva) [inline core/service/transaction/delegate.ts]: Address is number now

File: core/service/transaction/delegate.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    async verifyUnconfirmed(trs: Transaction<IAssetDelegate>, sender: Account): Promise<Response<void>> {
+        return new Response();
+    }
+
+    async verify(trs: Transaction<IAssetDelegate>, sender: Account): Promise<Response<void>> {
+        const errors = [];
+
+        if (trs.recipientAddress && config.constants.VERIFY_DELEGATE_TRS_RECIPIENT) {
```
Comment (axelsheva) [inline core/service/transaction/delegate.ts]: Please, remove config.constants.VERIFY* in all code

File: core/service/transaction/delegate.ts (hunk 1 above)
Comment (alex-arugon) [reply core/service/transaction/delegate.ts]: It's required in model and is checked during verify, so the field is protected.

File: core/repository/delegate.ts [hunk 4]
```diff
@@ -59,6 +59,26 @@ class DelegateRepository {
     public forgingStatus(publicKey: string) {
 
     }
+
+    public getByUsername(username: string): Delegate {
```
Comment (Verbalist) [inline core/repository/delegate.ts]: rewrite to get from acc and use link in acc

File: core/service/transaction/delegate.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+            asset.username = asset.username.toLowerCase().trim();
+        }
+        return asset;
+    }
+
+    getBytes(trs: Transaction<IAssetDelegate>): Buffer {
+        return Buffer.from(trs.asset.username, 'utf8');
+    }
+
+    async verifyUnconfirmed(trs: Transaction<IAssetDelegate>, sender: Account): Promise<Response<void>> {
```
Comment (Verbalist) [inline core/service/transaction/delegate.ts]: sync

File: core/service/transaction/delegate.ts [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        return new Response();
+    }
+
+    async verify(trs: Transaction<IAssetDelegate>, sender: Account): Promise<Response<void>> {
```
Comment (Verbalist) [inline core/service/transaction/delegate.ts]: sync

File: core/service/transaction/delegate.ts [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    async applyUnconfirmed(trs: Transaction<IAssetDelegate>, sender: Account): Promise<Response<void>> {
+        const newDelegate: Delegate = DelegateRepo.add(sender, {
+            username: trs.asset.username,
+            url: trs.asset.url
+        });
+        AccountRepo.attachDelegate(sender, newDelegate);
+        return new Response<void>();
+    }
+
+    async undoUnconfirmed(trs: Transaction<IAssetDelegate>, sender: Account): Promise<Response<void>> {
```
Comment (Verbalist) [inline core/service/transaction/delegate.ts]: sync

File: core/service/transaction/delegate.ts [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+    calculateFee(trs: Transaction<IAssetDelegate>, sender: Account): number {
+        return config.constants.fees.delegate;
+    }
+
+    calculateUndoUnconfirmed(trs: Transaction<IAssetDelegate>, sender: Account): void {
+        sender.delegate = null;
+    }
+
+    async applyUnconfirmed(trs: Transaction<IAssetDelegate>, sender: Account): Promise<Response<void>> {
```
Comment (Verbalist) [inline core/service/transaction/delegate.ts]: sync

File: core/repository/delegate.ts (hunk 4 above)
Comment (Verbalist) [reply core/repository/delegate.ts]: useless for core

File: core/repository/delegate.ts [hunk 9]
```diff
@@ -59,6 +59,26 @@ class DelegateRepository {
...
+        const iterations: number = this.memoryDelegates.length;
+        for (let i = 0; i < iterations; i++) {
+            if (this.memoryDelegates[i].username === username) {
+                return this.memoryDelegates[i];
+            }
+        }
+        return null;
+    }
+
+    public deleteByUsername(username: string): void {
```
Comment (Verbalist) [inline core/repository/delegate.ts]: same

**Pull request merged**